src/hal/lib/hal_ring.c

```diff
@@ -346,6 +346,10 @@ hal_plug_t *halg_plug_new(const int use_hal_mutex,
 	else
 	    ring = halg_find_object_by_id(0, HAL_RING, args->ring_id).ring;
 
+	// at this point the ring descriptor must be valid
+	if (ring == NULL)
+	    goto FAIL;
+
```

Comment:
True. Although, `goto` has some applications in C.

However, there are no reasons to use `goto` in C++ source code as done in other places. That's why there is try-catch.